plugins/cache-ehcache/src/main/java/org/opensearch/cache/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cache.keystore;
+
+import org.opensearch.common.cache.keystore.KeyLookupStore;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.core.common.unit.ByteSizeValue;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * This class implements KeyLookupStore using a roaring bitmap with a modulo applied to values.
+ * The modulo increases the density of values, which makes RBMs more memory-efficient. The recommended modulo is ~2^28.
+ * It also maintains a hash set of values which have had collisions. Values which haven't had collisions can be
+ * safely removed from the store. The fraction of collided values should be low,
+ * about 0.5% for a store with 10^7 values and a modulo of 2^28.
+ * The store estimates its memory footprint and will stop adding more values once it reaches its memory cap.
```

Comment:
Currently they just stay in memory, and the disk cache always goes to disk to look for a key (since it could get a false negative from a full keystore). We could also change it so the values are deleted to free memory, but the keystore is still marked as being full.

Reply:
> the disk cache always goes to disk to look for a key

This can cause frequent disk access if only the keys not present in RBM are accessed. We can have some form of eviction policy on the RBM to ensure this doesn't happen.